frame/referenda/src/lib.rs

```diff
@@ -579,6 +606,44 @@ pub mod pallet {
 				};
 			Ok(Some(branch.weight::<T, I>()).into())
 		}
+
+		/// Set the metadata to an ongoing referendum.
+		///
+		/// - `origin`: Must be `Signed`, and the creator of the referendum.
+		/// - `index`: The index of the referendum to add metadata for.
+		/// - `hash`: The preimage hash of an on-chain stored preimage.
+		#[pallet::weight(T::WeightInfo::set_metadata())]
+		pub fn set_metadata(
+			origin: OriginFor<T>,
+			index: ReferendumIndex,
+			hash: PreimageHash,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let status = Self::ensure_ongoing(index)?;
+			ensure!(status.submission_deposit.who == who, Error::<T, I>::NoPermission);
+			ensure!(T::Preimages::len(&hash).is_some(), Error::<T, I>::BadMetadata);
+			T::Preimages::request(&hash);
+			MetadataOf::<T, I>::insert(index, hash);
+			Self::deposit_event(Event::<T, I>::MetadataSet { index, hash });
+			Ok(())
+		}
+
+		/// Clear the referendum metadata.
+		///
+		/// - `origin`: Must be `Signed` or `Root`. If the referendum is ongoing, it must also be
+		///   the creator of the referendum.
+		/// - `index`: The index of the ongoing referendum to clear metadata for.
+		#[pallet::weight(T::WeightInfo::clear_metadata())]
+		pub fn clear_metadata(origin: OriginFor<T>, index: ReferendumIndex) -> DispatchResult {
+			let maybe_who = ensure_signed_or_root(origin)?;
+			if let Some(who) = maybe_who {
```

Comment:
If this is None then we proceed to `do_clear_metadata`, is that correct?

Reply:
Yep, if none, its Root

Reply:
same comment about using an option and just having a `set` api